core/client/tracker/tracker.go

```diff
@@ -0,0 +1,113 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package tracker verifies consistency proofs
+package tracker
+
+import (
+	"sync"
+
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/proto"
+	"github.com/google/trillian/types"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	pb "github.com/google/keytransparency/core/api/v1/keytransparency_go_proto"
+	tpb "github.com/google/trillian"
+)
+
+// UpdateTrustedPredicate return a bool indicating whether the local reference
+// for the latest SignedLogRoot should be updated.
+type UpdateTrustedPredicate func(cntRoot, newRoot types.LogRootV1) bool
+
+// LogRootVerifier verifies a Trillian Log Root.
+type LogRootVerifier interface {
+	VerifyRoot(trusted *types.LogRootV1, newRoot *tpb.SignedLogRoot, consistency [][]byte) (*types.LogRootV1, error)
+}
+
+// LogTracker keeps a continuous series of consistent log roots.
+type LogTracker struct {
+	trusted       types.LogRootV1
+	v             LogRootVerifier
+	updateTrusted UpdateTrustedPredicate
+	mu            sync.RWMutex
+}
+
+// New creates a log tracker from no trusted root.
+func New(lv LogRootVerifier) *LogTracker {
+	return NewFromSaved(lv, types.LogRootV1{})
+}
+
+// NewFromSaved creates a log tracker from a previously saved trusted root.
+func NewFromSaved(lv LogRootVerifier, lr types.LogRootV1) *LogTracker {
+	return &LogTracker{v: lv, trusted: lr, updateTrusted: isNewer}
+}
+
+// LastVerifiedLogRoot retrieves the tree size of the latest log root.
+func (l *LogTracker) LastVerifiedLogRoot() *pb.LogRootRequest {
+	l.mu.RLock()
+	defer l.mu.RUnlock()
+	return l.logRootRequest()
+}
+
+func (l *LogTracker) logRootRequest() *pb.LogRootRequest {
+	return &pb.LogRootRequest{
+		TreeSize: int64(l.trusted.TreeSize),
+		RootHash: l.trusted.RootHash,
+	}
+}
+
+// VerifyLogRoot verifies root and updates the trusted root if it is newer.
+// req must be equal to the most recent value from LastVerifiedLogRoot().
+// If two clients race to VerifyLogRoot at the same time, if one of them updates the root, the other will fail.
+func (l *LogTracker) VerifyLogRoot(req *pb.LogRootRequest, root *pb.LogRoot) (*types.LogRootV1, error) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	if want := l.logRootRequest(); !proto.Equal(req, want) {
+		glog.Warningf("logtracker: unexpected logRootRequest: %v, want %v", req, want)
+		return nil, status.Errorf(codes.InvalidArgument, "out of order VerifyLogRoot(%v, _), want %v", req, want)
+	}
+
+	logRoot, err := l.v.VerifyRoot(&l.trusted,
+		root.GetLogRoot(),
+		root.GetLogConsistency())
+	if err != nil {
+		return nil, err
+	}
+	if l.updateTrusted(l.trusted, *logRoot) {
+		l.trusted = *logRoot
+		glog.Infof("Trusted root updated to TreeSize %v", l.trusted.TreeSize)
+	}
+	return logRoot, nil
+}
+
+// SetUpdatePredicate allows tests to have finer grained control over when
+// the trusted root is updated.
```

Comment:
Add something like "Must be called before any other method."?

Reply:
It can be set at any time, but I'll suggest setting it before other calls.